### PR TITLE
Feature/cu jq3kzt screen resolution setting script

### PR DESCRIPTION
**Pull request by devildrake**
Established two scripts so that the system worked:

ScreenResolutionSetter -> Used to get OnClicked events on increase/decrease buttons
ScreenResolutionConfirmer -> Used to get OnClicked events on Confirmation button, as well as keeping track of the information sent to the Tesseract engine methods.

Had to make updates on ModuleRender and ModuleWindow:
- Module Render now receives SCREEN_RESIZED TESSERACTEVENT
- Module Window now  receives SCREEN_RESIZED TESSERACTEVENT (uses a bool to actually do the update due to SDL issues)
- Module Window now holds an int for the selected ResolutionPreset, as well as the enum class RESOLUTION_PRESET

An example implementation can be found on prefab named ScreenSizeObj.prefab

Tested both on Release and Game.

https://user-images.githubusercontent.com/23354524/117956566-74cbce00-b319-11eb-9dd2-11755b48ae66.mp4

### CODE REVIEW

File: Game/Assets/Scripts/ScreenResolutionSetter.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+void ScreenResolutionSetter::Start() {
+	GameObject* textObj = GameplaySystems::GetGameObject(textObjectID);
+
+	if (textObj) {
+		text = textObj->GetComponent<ComponentText>();
+	}
+
+	UpdateText();
+	screenResolutionChangeConfirmationWasRequested = false;
```
Comment (magalenyo) [inline Game/Assets/Scripts/ScreenResolutionSetter.cpp:28]: Lo estás inicializando en el Applier también?

File: Project/Source/Components/UI/ComponentCanvas.cpp [hunk 2]
```diff
@@ -57,13 +58,15 @@ float ComponentCanvas::GetScreenFactor() {
 
 void ComponentCanvas::RecalculateSizeAndScreenFactor() {
 	if (dirty) {
+		LOG("RECALCULATED");
```
Comment (magalenyo) [inline Project/Source/Components/UI/ComponentCanvas.cpp:61]: Remove debug log

File: Project/Source/Components/UI/ComponentCanvas.cpp [hunk 3]
```diff
@@ -57,13 +58,15 @@ float ComponentCanvas::GetScreenFactor() {
...
 		if (App->userInterface->IsUsing2D()) {
 			size = App->renderer->GetViewportSize();
 		} else {
 			size = GetScreenReferenceSize() * SCENE_SCREEN_FACTOR;
 		}
 		float2 factor = size.Div(screenReferenceSize);
 		screenFactor = factor.x < factor.y ? factor.x : factor.y;
+		dirty = false;
```
Comment (magalenyo) [inline Project/Source/Components/UI/ComponentCanvas.cpp:69]: I think it's this line that introduces a new bug!! The bug is when you load a new scene, everything is downscaled until you resize. Either remove it or fix it

File: Project/Source/GameplaySystems.cpp [hunk 4]
```diff
@@ -207,8 +207,23 @@ float Screen::GetScreenHeight() {
 	return static_cast<float>(App->window->GetHeight());
 }
 
+void Screen::SetResolution(int width_, int height_) {
+	App->window->SetSize(width_, height_);
```
Comment (magalenyo) [inline Project/Source/GameplaySystems.cpp:211]: Does Window->SetSize automatically triggers the viewportResized?

File: Project/Source/GameplaySystems.h [hunk 5]
```diff
@@ -391,6 +391,19 @@ namespace Input {
 namespace Screen {
 	TESSERACT_ENGINE_API float GetScreenWitdh();
 	TESSERACT_ENGINE_API float GetScreenHeight();
+	TESSERACT_ENGINE_API void SetResolution(int width_, int height_);
+	TESSERACT_ENGINE_API float2 GetResolution();
+
+	enum RESOLUTION_PRESET {
+		m_720x480 = 0,
```
Comment (magalenyo) [inline Project/Source/GameplaySystems.h:398]: You don't need the = 0.

File: Project/Source/Modules/ModuleRender.cpp [hunk 6]
```diff
@@ -252,6 +258,18 @@ void ModuleRender::ViewportResized(int width, int height) {
 	viewportUpdated = true;
 }
 
+void ModuleRender::ReceiveEvent(TesseractEvent& e) {
+	switch (e.type) {
+	case TesseractEventType::SCREEN_RESIZED:
+#if GAME
+		ViewportResized(App->window->GetWidth(), App->window->GetHeight());
```
Comment (magalenyo) [inline Project/Source/Modules/ModuleRender.cpp:265]: Not too sure about this. Compare with ViewportResized and make sure every ViewportResized is called.

File: Project/Source/Modules/ModuleScene.cpp [hunk 7]
```diff
@@ -93,7 +93,7 @@ bool ModuleScene::Start() {
 
 #if GAME
 	App->events->AddEvent(TesseractEventType::PRESSED_PLAY);
-	SceneImporter::LoadScene("Assets/Scenes/StartScene.scene");
+	SceneImporter::LoadScene("Assets/Scenes/Level1-34.scene");
```
Comment (magalenyo) [inline Project/Source/Modules/ModuleScene.cpp:96]: Undo this

File: Project/Source/Modules/ModuleWindow.h [hunk 8]
```diff
@@ -14,6 +14,13 @@ enum class WindowMode {
 	FULLSCREEN_DESKTOP
 };
 
+enum class RESOLUTION_PRESET {
+	m_720x480 = 0,
```
Comment (magalenyo) [inline Project/Source/Modules/ModuleWindow.h:18]: You don't need the = 0

File: Project/Source/Modules/ModuleWindow.h [hunk 9]
```diff
@@ -47,6 +56,8 @@ class ModuleWindow : public Module {
 	std::vector<SDL_DisplayMode> displayModes;
 
 private:
+	int currentResolutionPreset = 2;
```
Comment (magalenyo) [inline Project/Source/Modules/ModuleWindow.h:59]: You could set the m_1920x1080 directly

**Review by magalenyo — CHANGES_REQUESTED** [review]
Please, review the bug

File: Project/Source/Components/UI/ComponentCanvas.cpp (hunk 3 above)
Comment (JaimeBea) [reply Project/Source/Components/UI/ComponentCanvas.cpp:69]: This should be right. Maybe there is a problem where Invalidate is not being called correctly.

**Review by JaimeBea — APPROVED** [review]

File: Project/Source/Components/UI/ComponentCanvas.cpp (hunk 3 above)
Comment (JaimeBea) [reply Project/Source/Components/UI/ComponentCanvas.cpp:69]: The bug is caused by the following function not calling Invalidate():
```
void ComponentCanvas::SetScreenReferenceSize(float2 screenReferenceSize_) {
	screenReferenceSize = screenReferenceSize_;
        Invalidate();
}
```

File: Game/Assets/Prefabs/ScreenSizeObj.prefab [hunk 10]
```diff
@@ -0,0 +1,751 @@
+{
+    "Root": {
+        "Name": "ScreenSizeObj",
+        "Active": true,
```
[the remark this answers is not in the document]
Comment (devildrake) [reply Game/Assets/Prefabs/ScreenSizeObj.prefab:4]: True

File: Project/Source/Modules/ModuleWindow.h (hunk 9 above)
Comment (devildrake) [reply Project/Source/Modules/ModuleWindow.h:59]: Done

File: Project/Source/Modules/ModuleWindow.h (hunk 8 above)
Comment (devildrake) [reply Project/Source/Modules/ModuleWindow.h:18]: Okay

File: Project/Source/Modules/ModuleScene.cpp (hunk 7 above)
Comment (devildrake) [reply Project/Source/Modules/ModuleScene.cpp:96]: Done

File: Game/Assets/Scripts/ScreenResolutionSetter.cpp (hunk 1 above)
Comment (devildrake) [reply Game/Assets/Scripts/ScreenResolutionSetter.cpp:28]: En el ScreenResolutionConfirmer no se inicializa

File: Project/Source/Components/UI/ComponentCanvas.cpp (hunk 2 above)
Comment (devildrake) [reply Project/Source/Components/UI/ComponentCanvas.cpp:61]: Yup

File: Project/Source/GameplaySystems.h (hunk 5 above)
Comment (devildrake) [reply Project/Source/GameplaySystems.h:398]: True

File: Project/Source/Components/UI/ComponentCanvas.cpp (hunk 3 above)
Comment (devildrake) [reply Project/Source/Components/UI/ComponentCanvas.cpp:66]: Done. Had to make some further adjustments (Modifying 2D internal now calls all required invalidates

**Review by Jordi-Gil — APPROVED** [review]

File: Project/Source/Modules/ModuleRender.cpp (hunk 6 above)
Comment (devildrake) [reply Project/Source/Modules/ModuleRender.cpp]: Refactored, now using events.

**Review by JaimeBea — APPROVED** [review]

File: Project/Source/GameplaySystems.cpp (hunk 4 above)
Comment (devildrake) [reply Project/Source/GameplaySystems.cpp:211]: Yes

**Review by jcarlos0305 — APPROVED** [review]

**Review by Jordi-Gil — APPROVED** [review]

**Review by magalenyo — APPROVED** [review]
Approved, gj!

File: Project/Source/Modules/ModuleScene.cpp (hunk 7 above)
Comment (FrancPS) [reply Project/Source/Modules/ModuleScene.cpp:96]: It's not undone!

**Review by FrancPS — APPROVED** [review]

**Pull request merged**